sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamMatchRel.java

```diff
@@ -0,0 +1,279 @@
+                schemaBuilder.addField(collectionSchema.getField(index));
+            }
+            Schema mySchema = schemaBuilder.build();
+
+            // partition according to the partition keys
+            PCollection<KV<Row, Row>> keyedUpstream = upstream
+                .apply(ParDo.of(new MapKeys(mySchema)));
+
+            // sort within each partition
+            PCollection<KV<Row, ArrayList<Row>>> orderedUpstream = keyedUpstream
+                .apply(Combine.<Row, Row, ArrayList<Row>>perKey(new SortParts(mySchema, orderKeys)));
+
+            return null;
+        }
+
+        private static class SortParts extends Combine.CombineFn<Row, ArrayList<Row>, ArrayList<Row>> {
+
+            private final Schema mySchema;
+            private final List<RelFieldCollation> orderKeys;
+
+            public SortParts(Schema mySchema, RelCollation orderKeys) {
+                this.mySchema = mySchema;
+                List<RelFieldCollation> revOrderKeys = orderKeys.getFieldCollations();
+                Collections.reverse(revOrderKeys);
+                this.orderKeys = revOrderKeys;
+            }
+
+            @Override
+            public ArrayList<Row> createAccumulator() {
+                return new ArrayList<Row>();
+            }
+
+            @Override
+            public ArrayList<Row> addInput(ArrayList<Row> Accum, Row inRow) {
+                Accum.add(inRow);
+                return Accum;
+            }
+
+            @Override
+            public ArrayList<Row> mergeAccumulators(Iterable<ArrayList<Row>> Accums) {
+                ArrayList<Row> aggAccum = new ArrayList<Row>();
+                for (ArrayList<Row> i : Accums) {
+                    aggAccum.addAll(i);
+                }
+                return aggAccum;
+            }
+
+            @Override
+            public ArrayList<Row> extractOutput(ArrayList<Row> rawRows) {
+                for (RelFieldCollation i : orderKeys) {
+                    int fIndex = i.getFieldIndex();
+                    RelFieldCollation.Direction dir = i.getDirection();
+                    if (dir == RelFieldCollation.Direction.ASCENDING) {
+                        Collections.sort(rawRows, new sortComparator(fIndex, true));
+                    }
+                }
+                return rawRows;
+            }
+
+            private class sortComparator implements Comparator<Row> {
+
+                private final int fIndex;
+                private final int inv;
+
+                public sortComparator(int fIndex, boolean inverse) {
+                    this.fIndex = fIndex;
+                    this.inv = inverse ? -1 : 1;
+                }
+
+                @Override
+                public int compare(Row o1, Row o2) {
+                    Schema.Field fd = mySchema.getField(fIndex);
+                    Schema.FieldType dtype = fd.getType();
+                    switch (dtype.getTypeName()) {
```

Comment:
There is a Row comparator: https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamSortRel.java#L373